runtime/server/migrations.go

```diff
@@ -37,40 +38,126 @@ func (s *Server) MigrateSingle(ctx context.Context, req *api.MigrateSingleReques
 	if err != nil {
 		return nil, status.Error(codes.Unknown, err.Error())
 	}
 	olap, _ := conn.OLAPStore()
 
-	// Ingest the source
-	err = olap.Ingest(ctx, source)
-	if err != nil {
-		return nil, status.Error(codes.Unknown, err.Error())
+	// Get existing object with name and check it's a source
+	existingObj, existingFound := catalog.FindObject(ctx, req.InstanceId, source.Name)
+	if existingFound && !req.CreateOrReplace {
+		return nil, status.Errorf(codes.InvalidArgument, "an existing object with name '%s' already exists (consider setting `create_or_replace=true`)", existingObj.Name)
+	}
+	if existingFound && existingObj.Type != drivers.CatalogObjectTypeSource {
+		return nil, status.Errorf(codes.InvalidArgument, "an object of type '%s' already exists with name '%s'", existingObj.Type, existingObj.Name)
+	}
+
+	// Get object to rename and check it's a valid rename op
+	var renameObj *drivers.CatalogObject
+	var renameFound bool
+	var renameAndReingest bool
+	if req.RenameFrom != "" {
+		// Check that we're not renaming to a name that's already taken
+		if existingFound {
+			return nil, status.Errorf(codes.InvalidArgument, "cannot rename '%s' to '%s' because a source with that name already exists", req.RenameFrom, source.Name)
+		}
+
+		// Get the object to rename
+		renameObj, renameFound = catalog.FindObject(ctx, req.InstanceId, strings.ToLower(req.RenameFrom))
+		if !renameFound {
+			return nil, status.Errorf(codes.InvalidArgument, "could not find existing object named '%s' to rename", req.RenameFrom)
+		}
+		if renameObj.Type != drivers.CatalogObjectTypeSource {
+			return nil, status.Errorf(codes.InvalidArgument, "cannot rename object '%s' because it is not a source", req.RenameFrom)
+		}
+
+		// Check whether the properties for the new object are different (i.e. whether to re-ingest or just rename)
+		renameSource, err := sqlToSource(renameObj.SQL)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "could not parse existing sql: %s", err.Error())
+		}
+		renameAndReingest = !source.PropertiesEquals(renameSource)
+	}
+
+	// Stop execution now if it's just a dry run
+	if req.DryRun {
+		return &api.MigrateSingleResponse{}, nil
 	}
 
-	// Save definition
-	obj = &drivers.CatalogObject{
+	// Create the object to save
+	newObj := &drivers.CatalogObject{
 		Name: source.Name,
 		Type: drivers.CatalogObjectTypeSource,
 		SQL:  req.Sql,
 	}
-	if found {
-		err = catalog.UpdateObject(ctx, req.InstanceId, obj)
-	} else {
-		err = catalog.CreateObject(ctx, req.InstanceId, obj)
-	}
-	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error: could not insert source into catalog: %s (warning: watch out for corruptions)", err.Error())
+
+	// We now have several cases to handle
+	if !existingFound && !renameFound {
+		// Just ingest and save object
+		err := olap.Ingest(ctx, source)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
```

Comment:
I think this and the rest of the `olap.Ingest()` errors should be `codes.Unknown`.

(Was just working on user-friendly error messages on the frontend side)

Reply:
Responded in Slack